pkg/services/pluginsintegration/clientmiddleware/query_data_error_source_middleware.go

```diff
@@ -0,0 +1,84 @@
+package clientmiddleware
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+// errQueryDataDownstreamError is an returned when a plugin QueryData request has
+// failed with at least one QueryData response having statusSource = "downstream".
+var errQueryDataDownstreamError = errutil.Internal(
+	"plugin.queryDataDownstreamError",
+	errutil.WithPublicMessage("Plugin QueryData downstream error"),
+	errutil.WithDownstream(),
+)
+
+// NewQueryDataErrorSourceMiddleware creates a new plugins.ClientMiddleware
+// that will wrap QueryDataResponse error source as errors returned by QueryData.
+// This is used by [MetricsMiddleware] and [LoggerMiddleware] to correctly identify
+// plugin downstream errors.
+func NewQueryDataErrorSourceMiddleware() plugins.ClientMiddleware {
+	return plugins.ClientMiddlewareFunc(func(next plugins.Client) plugins.Client {
+		return &QueryDataErrorSourceMiddleware{
+			next: next,
+		}
+	})
+}
+
+type QueryDataErrorSourceMiddleware struct {
+	next plugins.Client
+}
+
+func (m *QueryDataErrorSourceMiddleware) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	resp, err := m.next.QueryData(ctx, req)
+
+	// Aggregate all errors whose ErrorSource is backend.ErrorSourceDownstream.
+	var queryDataResponseErr error
+	for _, r := range resp.Responses {
+		if r.Error == nil {
+			continue
+		}
+		if r.ErrorSource == backend.ErrorSourceDownstream {
+			queryDataResponseErr = errors.Join(queryDataResponseErr, r.Error)
+		} else {
+			// Other error source, this has higher priority, return the original
+			// error immediately.
+			return resp, err
+		}
+	}
+	if queryDataResponseErr != nil {
+		// If we have at least one downstream error, wrap the downstream errors and the original one in
+		// a errQueryDataDownstreamError.
+		return resp, errQueryDataDownstreamError.Errorf("query data downstream error: %w: %w", queryDataResponseErr, err)
```

Comment:
FYI I'm not sure we can do like this. Wouldn't an error response be returned from /api/ds/query then? Or is this just a way to signal to the logger and metrics middlewares what the status source is? To simplify, can we figure this out just for metrics to start with and then revisit this for logs later?

Reply:
Yes, exactly. the plan was to unwrap the error from `errQueryDataDownstreamError` before returning it, so `/api/ds/query` does not return an error and this error base is pretty much only used internally. However I'm still not sure if there's a cleaner way, so I am still experimenting with it

Reply:
As I said, would suggest handling this is in a follow up and only try solve the metrics part now to limit the scope. WDYT?

Reply:
Makes sense, I will leave logs for a follow-up PR